forc-plugins/forc-debug/src/server/state.rs

```diff
@@ -58,24 +59,55 @@ impl ServerState {
     pub fn vm_pc_to_source_location(
         &self,
         pc: Instruction,
-    ) -> Result<(&PathBuf, i64), AdapterError> {
-        // Try to find the source location by looking for the program counter in the source map.
-        self.source_map
+    ) -> Result<(PathBuf, i64), AdapterError> {
+        // Convert PC to instruction index (divide by 4 for byte offset)
+        let instruction_idx = (pc / 4) as usize;
+        if let Some((path, range)) = self.source_map.addr_to_span(instruction_idx) {
+            Ok((path, range.start.line as i64))
+        } else {
+            Err(AdapterError::MissingSourceMap { pc })
+        }
+    }
+
+    /// Updates the breakpoints in the VM for all remaining [TestExecutor]s.
+    pub(crate) fn update_vm_breakpoints(&mut self) {
+        if !self.breakpoints_need_update {
+            return;
+        }
+
+        // Convert breakpoints to instruction offsets using the source map
+        let opcode_indexes = self
+            .breakpoints
             .iter()
-            .find_map(|(source_path, source_map)| {
-                for (&line, instructions) in source_map {
-                    // Divide by 4 to get the opcode offset rather than the program counter offset.
-                    let instruction_offset = pc / 4;
-                    if instructions
-                        .iter()
-                        .any(|instruction| instruction_offset == *instruction)
-                    {
-                        return Some((source_path, line));
-                    }
-                }
-                None
-            })
-            .ok_or(AdapterError::MissingSourceMap { pc })
+            .flat_map(|(source_path, breakpoints)| {
+                breakpoints
+                    .iter()
+                    .filter_map(|bp| {
+                        bp.line.and_then(|line| {
+                            // Find any instruction that maps to this line in the source map
+                            self.source_map.map.iter().find_map(|(pc, _)| {
+                                self.source_map
+                                    .addr_to_span(*pc)
+                                    .filter(|(path, range)| {
+                                        path == source_path && range.start.line as i64 == line
+                                    })
+                                    .map(|_| pc)
+                            })
+                        })
+                    })
+                    .collect::<Vec<_>>()
+            });
+
+        // Set breakpoints in the VM
+        self.executors.iter_mut().for_each(|executor| {
+            // TODO: use `overwrite_breakpoints` when released
```

Comment:
Sorry I am missing some context around here. Is this going to be released from VM side?

Reply:
Yes, this has already been released: https://github.com/FuelLabs/fuel-vm/blob/b2bce26dec4265596157c7bf7b1c42a3a80d783d/fuel-vm/src/interpreter/debug.rs#L30

It's a connivence method so we can overwrite all of the breakpoints in one call rather than iterating. So lines 104-107 could become:

```
            let bps = opcode_indexes.clone().map(|&opcode_index| {
                fuel_vm::state::Breakpoint::script(opcode_index as u64)
            });
            executor.interpreter.overwrite_breakpoints(bps);
```